src/Silicon.cpp

```diff
@@ -825,70 +1082,252 @@ namespace galsim {
         _delta.resize(b);
         _delta.setZero();
+
+
+        // Map data to GPU
+        double* deltaData = _delta.getData();
+        int imageDataSize = _delta.getMaxPtr() - _delta.getData();
+
+        T* targetData = target.getData();
+        
+        int pixelInnerBoundsSize = _pixelInnerBounds.size();
+
+        int hbpSize = _horizontalBoundaryPoints.size();
+        int vbpSize = _verticalBoundaryPoints.size();
+
+        int hdSize = _horizontalDistortions.size();
+        int vdSize = _verticalDistortions.size();
+        
+        double* abs_length_table_data = _abs_length_table_GPU.data();
+        
+        int emptypolySize = _emptypoly.size();
+        Position<double>* emptypolyData = _emptypolyGPU.data();
+        
+        int nxny = nx * ny;
+	_changed.reset(new bool[nxny]);
+        bool* changedData = _changed.get();
+        for (int i = 0; i < nxny; i++) changedData[i] = false;
+
+        Bounds<double>* pixelInnerBoundsData = _pixelInnerBounds.data();
+        Bounds<double>* pixelOuterBoundsData = _pixelOuterBounds.data();
+
+        Position<float>* horizontalBoundaryPointsData = _horizontalBoundaryPoints.data();
+        Position<float>* verticalBoundaryPointsData = _verticalBoundaryPoints.data();
+        Position<float>* horizontalDistortionsData = _horizontalDistortions.data();
+        Position<float>* verticalDistortionsData = _verticalDistortions.data();
+
+#ifdef GALSIM_USE_GPU
+#pragma omp target enter data map(to: this[:1], deltaData[0:imageDataSize], targetData[0:imageDataSize], pixelInnerBoundsData[0:pixelInnerBoundsSize], pixelOuterBoundsData[0:pixelInnerBoundsSize], horizontalBoundaryPointsData[0:hbpSize], verticalBoundaryPointsData[0:vbpSize], abs_length_table_data[0:_abs_length_size], emptypolyData[0:emptypolySize], horizontalDistortionsData[0:hdSize], verticalDistortionsData[0:vdSize], changedData[0:nxny])
+#endif
+
+        // Start with the correct distortions for the initial image as it is already
+        dbg<<"Initial updatePixelDistortions\n";
+        updatePixelDistortions(target);
     }
 
+    void Silicon::finalize()
+    {
+#ifdef GALSIM_USE_GPU
+        Bounds<double>* pixelInnerBoundsData = _pixelInnerBounds.data();
+        Bounds<double>* pixelOuterBoundsData = _pixelOuterBounds.data();
+
+        Position<float>* horizontalBoundaryPointsData = _horizontalBoundaryPoints.data();
+        Position<float>* verticalBoundaryPointsData = _verticalBoundaryPoints.data();
+        Position<float>* horizontalDistortionsData = _horizontalDistortions.data();
+        Position<float>* verticalDistortionsData = _verticalDistortions.data();
+
+        double* abs_length_table_data = _abs_length_table_GPU.data();
+        
+        Bounds<int> b = _delta.getBounds();
+        const int nx = b.getXMax() - b.getXMin() + 1;
+        const int ny = b.getYMax() - b.getYMin() + 1;
+        int nxny = nx * ny;
+
+        int pixelInnerBoundsSize = _pixelInnerBounds.size();
```

Comment:
Maybe just call this `pixelBoundsSize` since it applies for both `_pixelInnerBounds` and `_pixelOuterBounds`.  (I was concerned for a moment reading `_pixelOuterBounds[0:pixelInnerBoundsSize]` below)

Reply:
Good point, I've renamed it to just `pixelBoundsSize`.